File: algorithms/mergesort.java

/**
 * Created by caden on 2017/5/9.
 */
public class mergeSort {
    public static void main(String[] args) {
        int[] seq = {1, 2, 3, 7, 5, 2, 3, 3, 1};
        sort(seq, 0, seq.length - 1);
        for (int i : seq) {
            System.out.println(i);
        }
    }

    public static void sort(int[] seq, int left, int right) {
        if (left < right) {
            int middle = (left + right) / 2;
            sort(seq, left, middle);
            sort(seq, middle + 1, right);
            merge(seq, left, right);
        }
    }

    public static void merge(int[] seq, int l, int r) {
        int mid = (l + r) / 2;
        int i = l;
        int j = mid + 1;
        int count = 0;
        int temp[] = new int[r - l + 1];
        while (i <= mid && j <= r) {
            if (seq[i] < seq[j]) {
                temp[count++] = seq[i++];
            } else {
                temp[count++] = seq[j++];
            }
        }
        while (i <= mid) {
            temp[count++] = seq[i++];
        }
        while (j <= r) {
            temp[count++] = seq[j++];
        }
        count = 0;
        while (l <= r) {
            seq[l++] = temp[count++];
        }
    }
}

File: algorithms/mergesort.py
#!/usr/bin/env python
#--coding:utf-8--
import math
import unittest
from testsortbase import algorithmTest

def merge(left, right):
	result = []
	n, m = 0, 0
	while n < len(left) and m < len(right):
		if left[n] <= right[m]:
			result.append(left[n])
			n += 1
		else:
			result.append(right[m])
			m += 1
	
	result += left[n:]
	result += right[m:]
	return result


def sort(seq):
	if len(seq) <= 1:
		return seq

	middle = len(seq) // 2
	left = sort(seq[:middle])
	right = sort(seq[middle:])
	return merge(left, right)

class TestMergeSort(algorithmTest):	
	def assertRange(self, seq):
		seqSorted = sort(seq)
		for x in range(len(seqSorted) - 1):
			self.assertTrue(seqSorted[x] <= seqSorted[x + 1])

if __name__ == '__main__':
	suite = unittest.TestLoader().loadTestsFromTestCase(TestMergeSort)
	unittest.TextTestRunner(verbosity=2).run(suite)
